### ISSUE TITLE
Right complexity for Hessian and HVPs

### ISSUE THREAD

**Issue by gdalle** [JuliaDiff/DifferentiationInterface.jl#84]
Let us consider a function `f` from `R^n` to `R`.

In what follows, I use
- "forward" for "backend that supports `O(1)` pushforwards"
- "reverse" for "backend that supports `O(1)` pullbacks

## Hessian-vector products

Sources:

- https://d2jud02ci9yv69.cloudfront.net/2024-05-07-bench-hvp-81/blog/bench-hvp/
- https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html

Complexity of one HVP:

| inner \ outer | forward | reverse |
|---|---|---|
| forward | `O(n)` | `O(1)` |
| reverse | `O(1)` | `O(1)` | 

HVP gives gradient as byproduct:

| inner \ outer | forward | reverse |
|---|---|---|
| forward | - | no |
| reverse | yes | no | 

## Hessian matrix

Complexity with "jacobian of gradient" approach: optimal if the gradient is computed in `O(1)` (reverse inner backend)

| inner \ outer | forward | reverse |
|---|---|---|
| forward | `O(n^2)` | `O(n^2)` |
| reverse | `O(n)` | `O(n)` | 

Complexity with iterated HVP approach:

| inner \ outer | forward | reverse |
|---|---|---|
| forward | `O(n^2)` | `O(n)` |
| reverse | `O(n)` | `O(n)` |

**Comment by avik-pal** [comment]
If you want to refer to a julia version for this, I have it here https://github.com/LuxDL/BatchedRoutines.jl/blob/1911ccf22699178e283c4b8fc40dbda7b6830323/src/chainrules.jl#L45 (ignore the batched_gradient naming :sweat: I needed something that I can export)

**Comment by gdalle** [comment]
How many HVPs are used at each solver iteration for optimization?

- One for each gradient?
- Many for each gradient?

cc @Vaibhavdixit02

**Comment by avik-pal** [comment]
If the loss function has a gradient term then just 1. But for a jacobian in
the gradient term we need to do N hvps, this can be chunked (I think) but I
haven't implemented it yet.

Note that I pretty much hardcode the inner gradient to be zygote and outer
jvp to be forwarddiff so it is not as general as you are targeting



Oops I thought this was a question to me. Github email truncated the cc part...

**Comment by gdalle** [comment]
Solved by #116
